401 Unauthorized exception marked as JsonParsingException

### Java API client version

8.5.2

### Java version

11

### Elasticsearch Version

8.5.2

### Problem description

When indexing a document the api, 401 unauthorized exception is marked as **JsonParsingException** which hides the actual error.

**Error message**
`Caused by: jakarta.json.stream.JsonParsingException: Jackson exception: Unexpected character ('<' (code 60)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')`

**Steps to reproduce.**
Setup a client with wrong password and index a document.

**Full stack trace**
`Caused by: jakarta.json.stream.JsonParsingException: Jackson exception: Unexpected character ('<' (code 60)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (ByteArrayInputStream); line: 1, column: 2]
	at co.elastic.clients.json.jackson.JacksonJsonpParser.convertException(JacksonJsonpParser.java:84) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.json.jackson.JacksonJsonpParser.fetchNextToken(JacksonJsonpParser.java:91) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.json.jackson.JacksonJsonpParser.next(JacksonJsonpParser.java:118) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:70) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.json.ObjectBuilderDeserializer.deserialize(ObjectBuilderDeserializer.java:79) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.json.DelegatingDeserializer$SameType.deserialize(DelegatingDeserializer.java:43) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.transport.rest_client.RestClientTransport.getHighLevelResponse(RestClientTransport.java:280) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.transport.rest_client.RestClientTransport.performRequest(RestClientTransport.java:148) ~[elasticsearch-java-8.5.2.jar:?]
	at co.elastic.clients.elasticsearch.ElasticsearchClient.index(ElasticsearchClient.java:962) ~[elasticsearch-java-8.5.2.jar:?]
`

--- a/java-client/src/main/java/co/elastic/clients/transport/TransportException.java
+++ b/java-client/src/main/java/co/elastic/clients/transport/TransportException.java
@@ -24,22 +24,32 @@ import java.io.IOException;
 
 public class TransportException extends IOException {
 
+    private final int statusCode;
     private final String endpointId;
 
-    public TransportException(String message, String endpointId) {
-        this(message, endpointId, null);
+    public TransportException(int statusCode, String message, String endpointId) {
+        this(statusCode, message, endpointId, null);
     }
 
-    public TransportException(String message, String endpointId, Throwable cause) {
-        super(endpointId == null ? message : "[" + endpointId + "] " + message, cause);
+    public TransportException(int statusCode, String message, String endpointId, Throwable cause) {
+        super("status: " + statusCode + ", " + (endpointId == null ? message : "[" + endpointId + "] " + message), cause);
+        this.statusCode = statusCode;
         this.endpointId = endpointId;
+    }
+
+    /**
+     * Status code returned by the http resquest
+     */
+    public int statusCode() {
+        return statusCode;
     }
 
     /**
      * Identifier of the API endpoint that caused the exception, if known.
      */
     @Nullable
-    String getEndpointId() {
+    public String endpointId() {
         return endpointId;
     }
+
 }

--- a/java-client/src/main/java/co/elastic/clients/transport/rest_client/RestClientTransport.java
+++ b/java-client/src/main/java/co/elastic/clients/transport/rest_client/RestClientTransport.java
@@ -36,6 +36,7 @@ import co.elastic.clients.transport.TransportOptions;
 import co.elastic.clients.util.ApiTypeHelper;
 import co.elastic.clients.util.BinaryData;
 import co.elastic.clients.util.MissingRequiredPropertyException;
+import jakarta.json.JsonException;
 import jakarta.json.stream.JsonGenerator;
 import jakarta.json.stream.JsonParser;
 import org.apache.http.HttpEntity;
@@ -85,7 +86,7 @@ public class RestClientTransport implements ElasticsearchTransport {
 
     /**
      * The {@code Future} implementation returned by async requests.
-     * It wraps the RestClient's cancellable and progagates cancellation.
+     * It wraps the RestClient's cancellable and propagates cancellation.
      */
     private static class RequestFuture<T> extends CompletableFuture<T> {
         private volatile Cancellable cancellable;
@@ -310,6 +311,7 @@ public class RestClientTransport implements ElasticsearchTransport {
                 JsonpDeserializer<ErrorT> errorDeserializer = endpoint.errorDeserializer(statusCode);
                 if (errorDeserializer == null) {
                     throw new TransportException(
+                        statusCode,
                         "Request failed with status code '" + statusCode + "'",
                         endpoint.id(), new ResponseException(clientResp)
                     );
@@ -318,6 +320,7 @@ public class RestClientTransport implements ElasticsearchTransport {
                 HttpEntity entity = clientResp.getEntity();
                 if (entity == null) {
                     throw new TransportException(
+                        statusCode,
                         "Expecting a response body, but none was sent",
                         endpoint.id(), new ResponseException(clientResp)
                     );
@@ -333,14 +336,17 @@ public class RestClientTransport implements ElasticsearchTransport {
                         // TODO: have the endpoint provide the exception constructor
                         throw new ElasticsearchException(endpoint.id(), (ErrorResponse) error);
                     }
-                } catch(MissingRequiredPropertyException errorEx) {
+                } catch(JsonException | MissingRequiredPropertyException errorEx) {
                     // Could not decode exception, try the response type
                     try {
                         ResponseT response = decodeResponse(statusCode, entity, clientResp, endpoint);
                         return response;
                     } catch(Exception respEx) {
                         // No better luck: throw the original error decoding exception
-                        throw new TransportException("Failed to decode error response", endpoint.id(), new ResponseException(clientResp));
+                        throw new TransportException(statusCode,
+                            "Failed to decode error response, check exception cause for additional details", endpoint.id(),
+                            new ResponseException(clientResp)
+                        );
                     }
                 }
             } else {
@@ -368,6 +374,7 @@ public class RestClientTransport implements ElasticsearchTransport {
                 // Expecting a body
                 if (entity == null) {
                     throw new TransportException(
+                        statusCode,
                         "Expecting a response body, but none was sent",
                         endpoint.id(), new ResponseException(clientResp)
                     );
@@ -395,7 +402,7 @@ public class RestClientTransport implements ElasticsearchTransport {
             return response;
 
         } else {
-            throw new TransportException("Unhandled endpoint type: '" + endpoint.getClass().getName() + "'", endpoint.id());
+            throw new TransportException(statusCode, "Unhandled endpoint type: '" + endpoint.getClass().getName() + "'", endpoint.id());
         }
     }
 
@@ -411,6 +418,7 @@ public class RestClientTransport implements ElasticsearchTransport {
                 return;
             }
             throw new TransportException(
+                clientResp.getStatusLine().getStatusCode(),
                 "Missing [X-Elastic-Product] header. Please check that you are connecting to an Elasticsearch "
                     + "instance, and that any networking filters are preserving that header.",
                 endpoint.id(),
@@ -419,7 +427,9 @@ public class RestClientTransport implements ElasticsearchTransport {
         }
 
         if (!"Elasticsearch".equals(header)) {
-            throw new TransportException("Invalid value '" + header + "' for 'X-Elastic-Product' header.",
+            throw new TransportException(
+                clientResp.getStatusLine().getStatusCode(),
+                "Invalid value '" + header + "' for 'X-Elastic-Product' header.",
                 endpoint.id(),
                 new ResponseException(clientResp)
             );

--- a/java-client/src/test/java/co/elastic/clients/transport/TransportTest.java
+++ b/java-client/src/test/java/co/elastic/clients/transport/TransportTest.java
@@ -1,0 +1,75 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package co.elastic.clients.transport;
+
+import co.elastic.clients.elasticsearch.ElasticsearchClient;
+import co.elastic.clients.json.jackson.JacksonJsonpMapper;
+import co.elastic.clients.transport.rest_client.RestClientTransport;
+import com.sun.net.httpserver.HttpServer;
+import org.apache.http.HttpHost;
+import org.elasticsearch.client.ResponseException;
+import org.elasticsearch.client.RestClient;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+
+public class TransportTest extends Assertions {
+
+    @Test
+    public void testXMLResponse() throws Exception {
+        HttpServer httpServer = HttpServer.create(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
+
+        httpServer.createContext("/_cat/indices", exchange -> {
+            exchange.sendResponseHeaders(401, 0);
+            OutputStream out = exchange.getResponseBody();
+            out.write(
+                "<?xml version=\"1.0\"?><error>Error</error>".getBytes(StandardCharsets.UTF_8)
+            );
+            out.close();
+        });
+
+        httpServer.start();
+        InetSocketAddress address = httpServer.getAddress();
+
+        RestClient restClient = RestClient
+            .builder(new HttpHost(address.getHostString(), address.getPort(), "http"))
+            .build();
+
+        ElasticsearchClient esClient = new ElasticsearchClient(new RestClientTransport(restClient, new JacksonJsonpMapper()));
+
+        TransportException ex = Assertions.assertThrows(
+            TransportException.class,
+            () -> esClient.cat().indices()
+        );
+
+        httpServer.stop(0);
+
+        assertEquals(401, ex.statusCode());
+        assertEquals("es/cat.indices", ex.endpointId());
+
+        // Cause is transport-dependent
+        ResponseException restException = (ResponseException) ex.getCause();
+        assertEquals(401, restException.getResponse().getStatusLine().getStatusCode());
+    }
+}

